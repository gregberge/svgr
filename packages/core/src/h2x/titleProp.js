import { JSXElement, JSXInterpolation } from 'h2x-plugin-jsx'

/**
 * Check if a classical title exist
 * @param {*} node - NodeElement
 * @returns {boolean}
 */
const hasTitle = node =>
  node.children.reduce((accumulation, value) => {
    if (value.name !== 'title') return accumulation
    if (value.children.some(e => e.text === 'title')) return accumulation
    return true
  }, false)

const titleProp = () => () => ({
  visitor: {
    JSXElement: {
      enter(path) {
        if (path.node.name === 'svg') {
          if (hasTitle(path.node)) {
            path.node.children = path.node.children.filter(element => {
              if (element.name !== 'title') return true
              if (element.children.some(e => e.text === 'title')) return true
              return false
            })
          }

          if (!path.node.children.some(children => children.name === 'title')) {
            const element = new JSXElement()
            const interpolation = new JSXInterpolation()
            interpolation.value = 'title'
            element.name = 'title'
            element.children.push(interpolation)
            path.node.children.unshift(element)
          }
        }
      },
    },
  },
})

export default titleProp
