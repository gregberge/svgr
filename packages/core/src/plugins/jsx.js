import unified from 'unified'
import parse from 'rehype-parse'
import vfile from 'vfile'
import hastToBabelAst from 'hast-util-to-babel-ast'
import { transformFromAstSync, createConfigItem } from '@babel/core'
import removeJSXAttribute from '../babelPlugins/removeJSXAttribute'
import removeJSXEmptyExpression from '../babelPlugins/removeJSXEmptyExpression'

export default (code, config, state = {}) => {
  const filePath = state.filePath || 'unknown'
  const hastTree = unified()
    .use(parse, {
      fragment: true,
      space: 'svg',
      emitParseErrors: true,
      duplicateAttribute: false,
    })
    .parse(vfile({ path: filePath, contents: code }))

  const babelTree = hastToBabelAst(hastTree)

  const { code: transformedCode } = transformFromAstSync(babelTree, code, {
    plugins: [
      createConfigItem([
        removeJSXAttribute,
        { attribute: 'xmlns' },
        'removeXmlns',
      ]),
      createConfigItem(removeJSXEmptyExpression),
    ],
    filename: filePath,
    babelrc: false,
    configFile: false,
  })

  return transformedCode
}
