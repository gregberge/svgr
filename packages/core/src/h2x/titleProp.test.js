import jsx from 'h2x-plugin-jsx'
import { transform } from 'h2x-core'
import titleProp from './titleProp'

describe('titleProp', () => {
  it('should add title element to svg', () => {
    const result = transform(
      `<svg>
          <path d="M51,37 L37,51" id="Shape"></path>
      </svg>
      `,
      { plugins: [jsx, titleProp()] },
    )

    expect(result).toMatchSnapshot()
  })

  it('should remove old title and add the new title', () => {
    const result = transform(
      `<svg>
          <title>Rectangle 5</title>
          <path d="M51,37 L37,51" id="Shape"></path>
      </svg>
      `,
      { plugins: [jsx, titleProp()] },
    )

    expect(result).toMatchSnapshot()
  })
})
