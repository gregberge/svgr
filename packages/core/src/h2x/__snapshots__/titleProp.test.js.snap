// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`titleProp should add title element to svg 1`] = `
"<svg>
  <title>
    {title}
  </title>
  <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
</svg>
"
`;

exports[`titleProp should remove old title and add the new title 1`] = `
"<svg>
  <title>
    {title}
  </title>
  <path d=\\"M51,37 L37,51\\" id=\\"Shape\\" />
</svg>
"
`;
